File: bfs.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX 100

struct Node
{
    int data;
    struct Node *next;
};

struct Graph
{
    int num_vertices;
    struct Node **adjlist;
    bool *visited;
};

struct Queue
{
    int items[MAX];
    int front;
    int rear;
};

void enqueue(struct Queue *queue, int value)
{
    if (queue->rear == MAX - 1)
    {
        return; // queue full
    }
    if (queue->front == -1)
    {
        queue->front = 0;
    }
    queue->rear++;
    queue->items[queue->rear] = value;
}

int dequeue(struct Queue *queue)
{
    if (queue->front == -1)
    {
        return -1;
    }

    int item = queue->items[queue->front];
    if (queue->front >= queue->rear)
    {
        queue->front = queue->rear = -1;
    }
    else
    {
        queue->front++;
    }
    return item;
}

bool isempty(struct Queue *queue)
{
    return (queue->front == -1);
}

struct Node *createnode(int data)
{
    struct Node *newnode = (struct Node *)malloc(sizeof(struct Node));
    newnode->data = data;
    newnode->next = NULL;
    return newnode;
}

struct Graph *creatgraph(int vertices)
{
    struct Graph *graph = (struct Graph *)malloc(sizeof(struct Graph));
    graph->num_vertices = vertices;

    graph->adjlist = (struct Node **)malloc(vertices * sizeof(struct Node *));

    for (int i = 0; i < graph->num_vertices; i++)
    {
        graph->adjlist[i] = NULL;
    }

    graph->visited = (bool *)malloc(vertices * sizeof(bool));
    for (int i = 0; i < vertices; i++)
    {
        graph->visited[i] = false;
    }
    return graph;
}

void addedge(struct Graph *graph, int src, int dest)
{
    struct Node *newnode = createnode(dest);
    newnode->next = graph->adjlist[src];
    graph->adjlist[src] = newnode;

    newnode = createnode(src);
    newnode->next = graph->adjlist[dest];
    graph->adjlist[dest] = newnode;
}

void BFS(struct Graph *graph, int startvertex)
{
    struct Queue queue;
    queue.front = queue.rear = -1;
    enqueue(&queue, startvertex);
    graph->visited[startvertex] = true;

    printf("BFS: ");
    while (!isempty(&queue))
    {
        int currentvertex = dequeue(&queue);
        printf("%d  ", currentvertex);
        struct Node *temp = graph->adjlist[currentvertex];
        while (temp != NULL)
        {
            int adjvertex = temp->data;
            if (!graph->visited[adjvertex])
            {
                enqueue(&queue, adjvertex);
                graph->visited[adjvertex] = true;
            }
            temp = temp->next;
        }
    }
    printf("\n");
}

void printgraph(struct Graph *graph)
{
    for (int i = 0; i < graph->num_vertices; i++)
    {
        struct Node *temp = graph->adjlist[i];
        printf("%d :", i);
        while (temp != NULL)
        {
            printf("%d ->", temp->data);
            temp = temp->next;
        }
        printf("NULL\n");
    }
}

int main(int argc, char const *argv[])
{
    int vertices, src, dest;
    printf("Enter number of vertices: ");
    scanf("%d", &vertices);

    struct Graph *graph = creatgraph(vertices);
    printf("Enter vertices src dest and -1 -1 to exit: ");

    while (1)
    {
        scanf("%d%d", &src, &dest);
        if (src == -1 && dest == -1)
        {
            break;
        }
        else if (src < 0 || src >= vertices || dest < 0 || dest >= vertices)
        {
            continue;
        }
        addedge(graph, src, dest);
    }

    printgraph(graph);
    printf("\n");
    BFS(graph, 0);

    return 0;
}

File: Python/bfs.py
class Graph:
    def __init__(self):
        self.adj_list = {}

    def add_edge(self, src, dest, weight):
        if src not in self.adj_list:
            self.adj_list[src] = []
        if dest not in self.adj_list:
            self.adj_list[dest] = []

        self.adj_list[src].append((dest, weight))
        # self.adj_list[dest].append(src)  #for undirected

    def print_graph(self):
        for src, dest_list in self.adj_list.items():
            print(f"Vertex {src}:", end='')
            for dest, weight in dest_list:
                print(f"({src}->({dest},{weight})", end='')
            print()

    def BFS(self, start):
        visited = {}
        queue = [start]

        while queue:
            vertex = queue.pop(0)
            if vertex not in visited:
                print(vertex, end=' ')
                visited[vertex] = True
                for neighbour,weight in self.adj_list[vertex]:
                    if neighbour not in visited:
                        queue.append(neighbour)


def main():
    graph = Graph()
    edges = int(input("Edges: "))

    for _ in range(edges):
        src, dest, weight = input("Enter src dest and weight").split()
        graph.add_edge(src, dest, int(weight))

    graph.print_graph()

    print("BFS")
    start = (input("Start vertex: "))
    graph.BFS(start)


if __name__ == '__main__':
    main()
